File: app/src/main/java/com/robotnec/budget/app/adapters/CategoriesAdapter.java
package com.robotnec.budget.app.adapters;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.robotnec.budget.R;
import com.robotnec.budget.app.util.TextIconUtils;
import com.robotnec.budget.core.domain.Category;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author zak <zak@swingpulse.com>
 */
public class CategoriesAdapter extends BaseAdapter<Category, CategoriesAdapter.ViewHolder> {

    private CategoryClickListener categoryClickListener;

    public CategoriesAdapter(Context context, CategoryClickListener listener) {
        super(context);
        categoryClickListener = listener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = layoutInflater.inflate(R.layout.item_category, parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        Category category = items.get(position);
        String name = category.getName();

        holder.textName.setText(name);

        holder.imageIcon.setImageDrawable(TextIconUtils.generate(name));

        holder.itemView.setOnClickListener(v -> {
            if (categoryClickListener != null) {
                categoryClickListener.onCategoryClick(items.get(holder.getAdapterPosition()));
            }
        });
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.text_category_name)
        TextView textName;

        @BindView(R.id.image_icon)
        ImageView imageIcon;

        ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    public interface CategoryClickListener {
        void onCategoryClick(Category category);
    }
}

File: app/src/main/java/com/robotnec/budget/app/adapters/CategoriesAdapter.kt
package com.robotnec.budget.app.adapters

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import com.robotnec.budget.R
import com.robotnec.budget.app.adapters.support.BaseAdapter
import com.robotnec.budget.app.util.TextIconUtils
import com.robotnec.budget.core.domain.Category
import kotlinx.android.synthetic.main.item_category.view.imageIcon
import kotlinx.android.synthetic.main.item_category.view.textCategoryName

/**
 * @author zak zak@swingpulse.com>
 */
class CategoriesAdapter(context: Context, private val categoryClickListener: CategoriesAdapter.CategoryClickListener)
    : BaseAdapter<Category, CategoriesAdapter.ViewHolder>(context) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemView = layoutInflater.inflate(R.layout.item_category, parent, false)
        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val (_, name) = items[position]

        holder.itemView.textCategoryName.text = name
        holder.itemView.imageIcon.setImageDrawable(TextIconUtils.generate(name))
        holder.itemView.setOnClickListener { v ->
            categoryClickListener.onCategoryClick(items[holder.adapterPosition])
        }
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    interface CategoryClickListener {
        fun onCategoryClick(category: Category)
    }
}
